src/sietch/distributed_dune.ml

```diff
@@ -61,23 +95,29 @@ let client t target =
            (Digest.to_string target))
   in
   let* uri = find_target target |> Async.return in
-  match Clients.find t.clients uri with
-  | Some client -> Async.Deferred.Result.return client
-  | None -> connect t uri
+  let+ client =
+    match Clients.find t.clients uri with
+    | Some client -> Async.Deferred.Result.return client
+    | None -> connect t uri
+  in
+  (client, uri)
+
+let client t target = client_and_uri t target >>| fst
 
 let block_get t hash =
   let* client = client t hash in
-  Async.Rpc.Rpc.dispatch_exn Dune_distributed_storage.Rpc.block_get client
+  let ( >>= ) = Async.( >>= ) in
+  Async.Rpc.Pipe_rpc.dispatch Dune_distributed_storage.Rpc.block_get client
     (Digest.to_string hash)
-  |> async_ok
+  >>= Dune_distributed_storage.Rpc.decode_block_get |> async_ok
 
-let block_has t hash =
+let _block_has t hash =
```

Comment:
Are `_block_has` and `_block_put` not needed any more? Shall we delete them then?